aesara/tensor/basic.py

```diff
@@ -1302,17 +1302,26 @@ def eye(n, m=None, k=0, dtype=None):
 
     Returns
     -------
-    ndarray of shape (N,M)
-        An array where all elements are equal to zero, except for the `k`-th
-        diagonal, whose values are equal to one.
-
+    aesara tensor of shape (N,M)
+        A symbolic tensor representing a matrix where all elements are equal to zero,
+        except for the `k`-th diagonal, whose values are equal to one.
     """
-    if dtype is None:
-        dtype = config.floatX
+
     if m is None:
         m = n
-    localop = Eye(dtype)
-    return localop(n, m, k)
+    if dtype is None:
+        dtype = aesara.config.floatX
+
+    n = as_tensor_variable(n)
+    m = as_tensor_variable(m)
+    k = as_tensor_variable(k)
+
+    i = switch(k >= 0, k, -k * m)
+    eye = zeros(n * m, dtype=dtype)
+    eye = aesara.tensor.subtensor.set_subtensor(eye[i :: m + 1], 1).reshape((n, m))
+    eye = aesara.tensor.subtensor.set_subtensor(eye[m - k :, :], 0)
```

Comment:
Do we need the last `set_subtensor` (i.e. that appears to fill with `0`s if we've already used `zeros`?  I think I might know why, but it would be nice to have only one `set_subtensor`.

Reply:
The second `set_subtenor` is to prevent the diagonal from "wrapping around" the matrix in cases when `m > n` or when `k != 0`. This one-liner almost works:

```
at.subtensor.set_subtensor(eye[i:(m - k)*m:m + 1], 1).reshape((n_at, m_at))
```

But it fails when `k < -n` or `k > n`. In these cases I guess an error could be raised (why is the user asking for an offset greater than the width of the matrix?), but numpy returns the zero matrix for this case and I wanted to be consistent.

Reply:
One way to match `numpy` and avoid the second `set_subtensor` would be to check the conditions where the zero matrix should be returned and use multiplication:

```python
    eye = at.subtensor.set_subtensor(eye[i:(m - k)*m:m + 1], 1).reshape((n, m))
    eye *= (1 - at.gt(k, m)) * (1 - at.lt(k, -n))
```

But this was about 5µs slower than the two `set_subtensor` method when I ran `%timeit`. Perhaps it's too cute?

Reply:
Here's an implementation that uses scalars and `Composite`:
```python
import numpy as np
import aesara
import aesara.tensor as at
from aesara.scalar import int64, switch, Composite


n = int64("n")
m = int64("m")
k = int64("k")

i = switch(k >= 0, k, -k * m)

eye_sc = at.zeros(n * m, dtype="float64")
mkm = (m - k) * m
eye_sc = at.subtensor.set_subtensor(eye_sc[i : mkm : m + 1], 1).reshape((n, m))

eye_sc_fn = aesara.function([n, m, k], [eye_sc])

aesara.dprint(eye_sc_fn)
# Reshape{2} [id A] 16
#  |IncSubtensor{InplaceSet;int64:int64:int64} [id B] 15
#  | |Reshape{2} [id C] 14
#  | | |IncSubtensor{InplaceSet;int64:int64:int64} [id D] 13
#  | | | |Alloc [id E] 11
#  | | | | |TensorConstant{0.0} [id F]
#  | | | | |TensorFromScalar [id G] 8
#  | | | |   |mul [id H] 2
#  | | | |     |n [id I]
#  | | | |     |m [id J]
#  | | | |TensorConstant{1} [id K]
#  | | | |Switch [id L] 12
#  | | | | |GE [id M] 6
#  | | | | | |k [id N]
#  | | | | | |ScalarConstant{0} [id O]
#  | | | | |k [id N]
#  | | | | |mul [id P] 10
#  | | | |   |neg [id Q] 5
#  | | | |   | |k [id N]
#  | | | |   |m [id J]
#  | | | |mul [id R] 9
#  | | | | |sub [id S] 4
#  | | | | | |m [id J]
#  | | | | | |k [id N]
#  | | | | |m [id J]
#  | | | |add [id T] 3
#  | | |   |m [id J]
#  | | |   |ScalarConstant{1} [id U]
#  | | |MakeVector{dtype='int64'} [id V] 7
#  | |   |TensorFromScalar [id W] 1
#  | |   | |n [id I]
#  | |   |TensorFromScalar [id X] 0
#  | |     |m [id J]
#  | |TensorConstant{1} [id K]
#  | |Switch [id L] 12
#  | |mul [id R] 9
#  | |add [id T] 3
#  |MakeVector{dtype='int64'} [id V] 7

%timeit eye_sc_fn(10, 10, 1)
# 66.5 µs ± 1.42 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

# Manually construct composite `Op`s for the scalar operations
i_comp_op = Composite([n, m, k], [i])
i_comp = i_comp_op(n, m, k)

mkm_comp_op = Composite([m, k], [mkm])
mkm_comp = mkm_comp_op(m, k)

eye_sc_comp = at.zeros(n * m, dtype="float64")
eye_sc_comp = at.subtensor.set_subtensor(eye_sc_comp[i_comp : mkm_comp : m + 1], 1).reshape((n, m))

eye_sc_comp_fn = aesara.function([n, m, k], [eye_sc_comp])

aesara.dprint(eye_sc_comp_fn)
# Reshape{2} [id A] 10
#  |IncSubtensor{InplaceSet;int64:int64:int64} [id B] 9
#  | |Alloc [id C] 8
#  | | |TensorConstant{0.0} [id D]
#  | | |TensorFromScalar [id E] 7
#  | |   |mul [id F] 6
#  | |     |n [id G]
#  | |     |m [id H]
#  | |TensorConstant{1} [id I]
#  | |Composite{Switch(GE(i2, 0), i2, ((-i2) * i1))} [id J] 5
#  | | |n [id G]
#  | | |m [id H]
#  | | |k [id K]
#  | |Composite{((i0 - i1) * i0)} [id L] 4
#  | | |m [id H]
#  | | |k [id K]
#  | |add [id M] 3
#  |   |m [id H]
#  |   |ScalarConstant{1} [id N]
#  |MakeVector{dtype='int64'} [id O] 2
#    |TensorFromScalar [id P] 1
#    | |n [id G]
#    |TensorFromScalar [id Q] 0
#      |m [id H]

%timeit eye_sc_comp_fn(10, 10, 1)
# 60.9 µs ± 2.53 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Reply:
Before I forget, using `inc_subtensor` with `set_instead_of_inc=True` and `ignore_duplicates=True` might be faster.

Reply:
Is it a bug/to-do that `ifelse` can't be used with `Composite`?

Reply:
> Is it a bug/to-do that `ifelse` can't be used with `Composite`?

I don't think so, but we should create one.  I think the idea would amount to making a scalar version of `IfElse` with an `Op.c_code` implementation, which will require some consideration.

Reply:
We also need an issue for a rewrite that replaces scalar-only sub-graphs with `Composite`s.

Reply:
Sorry, `Switch` is quite literally the scalar version of `IfElse`&mdash;at least in terms of its `Op.c_code` implementation.    If anything, we should make invocations of `ifelse` return `Switch`es when the arguments are `ScalarType`ed.

Reply:
Here's the issue for `ifelse`: https://github.com/aesara-devs/aesara/issues/1226.

Reply:
Also, there's no scalar `Op`  for `max` and `min`?

Reply:
Disregard, obviously this can be done using scalar `switch`.